### PR TITLE
In commandline restore, in multi targeting scenarios, read the target framework details from the inner build

### LINKED ISSUE #9869: In command line restore, with multi targeted projects, NuGet should read the target framework related information from the inner build
Child of: https://github.com/NuGet/Home/issues/9756

Currently in command line scenarios NuGet reads the target framework information in the outer build.
That is incorrect because the the contract is that we should depend on on TargetFrameworkIdentifier and the like properties.
We do that here: https://github.com/NuGet/NuGet.Client/blob/bd482514a87a0085f8eb8603842ee1cb3ca5f277/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L500-L526

This task covers, the non experimental restore changes in command line scenarios, read the framework in the inner build (this should allow us to read OS Version for example).

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9869
Regression: No 
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: 
Currently in command line scenarios NuGet reads the target framework information in the outer build. 
That is incorrect because the the contract is that we should depend on on TargetFrameworkIdentifier and the like properties. 
We do that here: https://github.com/NuGet/NuGet.Client/blob/bd482514a87a0085f8eb8603842ee1cb3ca5f277/src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets#L500-L526

This issue covers  commandline restore (nuget.exe, msbuild.exe dotnet.exe).

Roughly some more important notes about the changes: 

* We *do not* parse the `TargetFramework` property any more.
* When grouping the dependencies, we group them over the TargetFramework alias. 
* Added a TargetAlias in each of the models containing pivots for dependencies such as `TargetFrameworkInformation` & `ProjectRestoreMetadataFrameworkInfo`. The first one contains the package refs, the latter one contains the project refs.
* When setting conditions, we'll retain the original value of `TargetFramework`. This is normal, but think it's worth calling it out. 
* We read the tfm information through a bunch of props. The item we define for this information will remain the one that was used for AssetTargetFallback (which was already being read from the inner build). 

* The target framework generation scenarios are hacky. Those changes are going to be replaced with whatever comes out of https://github.com/NuGet/NuGet.Client/pull/3549 + https://github.com/nuget/home/issues/9871.
* When I generate the dg spec for the NuGet.sln, there are 2 differences:
  * Each frameworks setting has a "targetAlias" value. 
  * net5.0 is getting parsed as `net5.0-windows7.0` because the SDK 5.0.100 P7 reports default TargetPlatformIdentifier and related.

**Note that this change is time critical. Any cosmetic suggestions, I'll be happy to address, but in a future PR**. 

## Testing/Validation

Tests Added: Yes  
Reason for not adding tests:  
Validation:

### CODE REVIEW

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:436]: why not an error?

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:944]: Do SDK style projects support TFMs with a profile? If so, we also need to pass through `TargetFrameworkProfile`

File: src/NuGet.Core/NuGet.Build.Tasks/WriteRestoreGraphTask.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks/WriteRestoreGraphTask.cs]: this file doesn't have any meaningful changes and can be reverted.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs]: use explicit type

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs]: use explicit type

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs]: use explicit type

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs]: explicit type

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs]: explicit type

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:795]: Maybe this will make more sense when I read tests, but is this because `dotnet tool install` creates a project with only a `TargetFrameworks` element, and the .NET SDK does not expand it into the 5 `TargetFramework*` properties?

`ProjectStyle.Unknown` also makes me think. Do we want the default to be to use NuGet's "short folder name", or the .NET SDK style 5 properties?  If we want the default to be NuGet's short folder style string, it seems strange to me why we're going to so much effort to implement aliases and the 5 properties.

File: src/NuGet.Core/NuGet.Commands/Utility/MSBuildProjectFrameworkUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/Utility/MSBuildProjectFrameworkUtility.cs]: indeed :P   This is a github reminder to fix before merging.

File: src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreReference.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreReference.cs]: This file has no meaningful changes and can be reverted.

File: src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt
Comment (zivkan) [inline src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt]: Does `GetTargetFramework` need to be public?  I don't see it used in this PR.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs]: remove trailing whitespace (your editor doesn't support [`.editorconfig`'s `trim_trailing_whitespace`](https://github.com/NuGet/NuGet.Client/blob/ad08bc710b3c4e142040f4efc860c003502aee3e/.editorconfig#L13) :(

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs]: remove trailing whitespace

**Review by zivkan — COMMENTED** [review]
Maybe I missed it, but I didn't see any tests where target platform identifier or version was set.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
Comment (zkat) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: aren't we supposed to keep supporting 4-part versions?

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs
Comment (zkat) [inline test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs:486]: what's this for?

**Review by zkat — COMMENTED** [review]
Most of this is over my head! But the tests look good and that's good enough for me! :P

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:944]: Technically we would get it out of the TargetFrameworkMoniker. 
It really depends on the utility that is written.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:436]: We want to be careful about duplicates. 

I'd be afraid to change this behavior now. While I don't see duplicates generated by the targets now, I think it's ok to be dup safe.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:795]: `DotnetCliTool` is *not* the `dotnet tool` gesture actually. It's the old project tools and those have a specific property that represents the tfm to restore against. Those are special.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:795]: ProjectStyle.Unknown is NuGet not enabled projects. 
When a project is not NuGet enabled, it can only be a legacy project, so reading the tfm the same we did in the old made sense to me. 

I can try to clean it up if you want me to, It was just what worked, and I didn't want to complicate it.

File: src/NuGet.Core/NuGet.Commands/Utility/MSBuildProjectFrameworkUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/Utility/MSBuildProjectFrameworkUtility.cs]: Yep, here I depend on the utility you are working on :D

File: src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt
Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt]: It's used in MSBuildRestoreUtility. 

I can just define an internal extension if you want me to. I don't necessarily know if we'll need it outside of commands...probably later if we try to allow duplicates.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs
Comment (nkolev92) [reply test/NuGet.Clients.Tests/NuGet.CommandLine.Test/RestoreNETCoreTest.cs]: It was my test infra that was lacking. I've reverted it.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs]: Fixed. It's ccode...I use VS and code interchangeably....VS is problematic when you want to make many changes across files.

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs
Comment (nkolev92) [reply test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestProjectContext.cs:486]: Our test infra pretends like we have an sdk based project...because we are not yet testing against one that can infer props...so I needed to do this to make the tests pass on 3.1.xxx. 

On another note...we should stop 3.1.xxx tests

**Review by nkolev92 — COMMENTED** [review]
@zivkan Addressed your feedback...one follow for you I think. 

@Zkat addressed your feedback.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs:436]: fwiw, this is equivalent to what we were doing before this change. The difference is that here we're deduping based on the alias, before we doing it based on the TFM.

File: src/NuGet.Core/NuGet.Commands/Utility/MSBuildProjectFrameworkUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/Utility/MSBuildProjectFrameworkUtility.cs]: Ok, utility here: https://github.com/NuGet/Home/issues/9871

**Review by zkat — APPROVED** [review]

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets:944]: Clarifying this, it depends on the utility that comes out of https://github.com/nuget/home/issues/9871.

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets]: trailing whitespace

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs]: whitespace formatting

File: src/NuGet.Core/NuGet.ProjectModel/PackageSpecExtensions.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.ProjectModel/PackageSpecExtensions.cs]: What's a scenario where a target alias is defined, but doesn't exist in the package spec?

I know NuGet has a history of silently ignoring errors, but should we be throwing a `KeyNotFoundException`, or something like that?  If the API caller requests a target alias that was not defined in the csproj, I think it should throw an exception.  But if it's possible that the targetAlias is defined in the csproj, but for some other reason is not in the packageSpec, then I'd find this more acceptable.

File: src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt
Comment (zivkan) [reply src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt]: It's not awful, so I won't block the PR on it, but from an API design point of view, I don't like it. If it exists because expecting callers to use `packageSpec.TargetFrameworks.SingleOrDefault(...)` is considered too hard, then we should probably consider a `packageSpec.TargetFrameworks[string]` indexer, rather than this helper extension.  Plus my comment on the implementation of the method itself, where its behaviour is not at all what I personally would expect (return value when requested key does not exist)

**Review by zivkan — APPROVED** [review]
just an idealistic (nitpick) concern about API design.

File: src/NuGet.Core/NuGet.ProjectModel/PackageSpecExtensions.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/PackageSpecExtensions.cs]: I think you are right. 
I'll remove the extension method, or rather I'll make it internal to NuGet.Commands and make it fail fast.

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:1506]: Does the restore code expect alias to always be set?

I think `TargetAlias` should be defaulted to `frameworkPropertyName`'s value up on line 1451, and overridden here if the json specified it.

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:1579]: Does the restore code expect alias to always be set?

I think `TargetAlias` should be defaulted to `jsonReader.Value`'s value up on line 1517, and overridden here if the json specified it.

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:1506]: > Does the restore code expect alias to always be set?

No, only when a project is multi targeted

Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:1506]: well, only when a project is SDK based :D

File: src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/JsonPackageSpecReader.cs:1579]: Not really. 

This is an already processed packagespec (as in, if it needed an alias, it would've been generated and written).
We shouldn't make assumptions like that here.

File: src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt
Comment (nkolev92) [reply src/NuGet.Core/NuGet.ProjectModel/PublicAPI.Unshipped.txt]: removed.

**Pull request merged**
